### PR TITLE
Persistent Scores, scrolling background

### PR DESCRIPTION
**Pull request by wildjames**
hey hey hey I'm just a lil guy.

Added a high score. This just gets saved to a file... as plain text... but we can hope no-one notices for now.

Also added a stock image infinitely scrolling background. There's room for improvement here in two aspects: The image is shit, and changing its scrolling speed causes a stutter. 

I know why the stutter happens though, but don't have any clean ideas on how to fix it. Shaders only have access to a variable TIME, which is the time (in seconds?) since the game started. The image is moved by a shader each frame, which translates the image left by TIME * scroll_speed. So, if we alter scroll_speed, the image jerks by some amount to its new position. Cant be arsed to fix it yet but a constant speed looks fine anyway, I think. Wouldn't be hard to calculate the offset in the main script, and pass the new rotation angle to the shader each frame, but it sounds ugly to me so I'll leave it till I get a better idea.

### CODE REVIEW

**Comment by wildjames** [comment]
Did some groundwork on the multiplayer. Only a first pass, needs some work. Builds fine on my mac but fails the tests due to a missing library?

Currently only tested with two players, and it works. Walls spawn on both and behave properly.

Walls are randomly generated - Jasper has the idea of simply synchronising the random seeds across players. Shouldn't be too hard to stick this in just before the game starts, but CBA tonight. A bit trickier will be handling player death - currently, the two instances are 100% independent of each other, just the characters appear on both screens.

File: export_presets.cfg [hunk 1]
```diff
@@ -24,7 +24,6 @@ binary_format/embed_pck=true
 custom_template/release=""
```
Comment (Jibbajabbafic) [inline export_presets.cfg]: Can you revert the changes to this file? It's been set up for the auto builds, so changes shouldn't be committed to it

Comment (wildjames) [reply export_presets.cfg]: Yep, sure.

**Comment by Jibbajabbafic** [comment]
Btw for future PRs it's much better to split up your changes, other wise they become a nightmare to review, so try to keep 1 feature per PR if possible

File: scripts/Connecting.gd [hunk 2]
```diff
@@ -0,0 +1,19 @@
+extends ColorRect
+
+const CONNECT_TEXT = ["Waiting for players...", "Connecting to server..."]
+
+
+func set_connect_type(host):
```
Comment (JordanPawlett) [inline scripts/Connecting.gd]: Seems confusing to name the signals method handler the same as the signal itself. The docs seem to suggest naming the handler `_on_<node_name>_<signal_name>`

File: scripts/HostButton.gd [hunk 3]
```diff
@@ -0,0 +1,17 @@
+extends Node
+
+signal set_connect_type
```
Comment (JordanPawlett) [inline scripts/HostButton.gd]: Currently `set_connect_type` seems to be expecting a boolean, this doesn't make too much sense to me. Maybe it would make more sense to name the signal `set_is_host`, as that seems to be what it's doing.

Comment (JordanPawlett) [inline scripts/HostButton.gd]: `set_connect_type` would be more expandable if it expected an enumerated type. This way you could easily add the ability for a spectator to join or different client types...

Not sure what the syntax is, but something like.
```
const CONNECTION_TYPES = {
    "host": 0,
    "player": 1,
    "spectator": 2
}
```

File: scenes/Lobby.tscn [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+[node name="Background" type="ColorRect" parent="."]
+margin_right = 1024.0
+margin_bottom = 600.0
+color = Color( 0.25098, 0, 0.380392, 1 )
+__meta__ = {
+"_edit_use_anchors_": false
+}
+
+[node name="Host" type="Button" parent="."]
+margin_left = 416.0
```
Comment (JordanPawlett) [inline scenes/Lobby.tscn]: Will all these positioning variables scale with the users window size?

File: scenes/Lobby.tscn [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+margin_top = 104.0
+margin_right = 608.0
```
Comment (JordanPawlett) [inline scenes/Lobby.tscn]: There's a lot of 'magic numbers' when defining the margins in this file. Can anyone of them be abstracted into a re-usable variable?
Are they based on anything specific? e.g. half a texture size, etc...

File: scripts/World.gd [hunk 6]
```diff
@@ -5,47 +5,108 @@ var height_range = 100
...
+	var saved = File.new()
+	if not saved.file_exists(high_score_fname):
+		return high_score
+	
+	saved.open(high_score_fname, File.READ)
+	while saved.get_position() < saved.get_len():
+		var data = parse_json(saved.get_line())
+		high_score = int(data['highscore'])
+		print("Loaded in save data")
+		print(data)
```
Comment (JordanPawlett) [inline scripts/World.gd]: I'd avoid printing stuff like this to the console, over time it will take up a lot of memory on the server

File: project.godot [hunk 7]
```diff
@@ -22,9 +22,17 @@ _global_script_class_icons={
 
 config/name="Flappy Race"
 config/description="Multiplayer flappy bird"
-run/main_scene="res://scenes/World.tscn"
+run/main_scene="res://scenes/Lobby.tscn"
 config/icon="res://icon.png"
 
+[autoload]
+
+Net="*res://scripts/network.gd"
```
Comment (JordanPawlett) [inline project.godot]: I'm obviously not familiar with the language but...
It looks like the Network instance is loaded as a singleton here, which wouldn't allow you to have more than 1 game active at any time. 

Seems more sensible to have a Controller singleton which would handle registering and unregistering a new instance of network each time a game is created

Comment (JordanPawlett) [reply project.godot]: You can probably ignore this comment. Just realised the instance of the game will be running on an individuals device, so it's probably fine being a singleton!

File: scripts/Connecting.gd [hunk 8]
```diff
@@ -0,0 +1,19 @@
+extends ColorRect
```
Comment (JordanPawlett) [inline scripts/Connecting.gd]: You may want to think about coupling these scripts files into logical groups and create sub-directories for them, otherwise /scripts is going to become very hard to navigate.

File: scenes/Lobby.tscn [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+margin_top = 320.0
+margin_right = 608.0
+margin_bottom = 344.0
+text = "Your IP: [Private IP]"
+align = 1
+valign = 1
+__meta__ = {
+"_edit_use_anchors_": false
+}
+[connection signal="pressed" from="Host" to="Host" method="host"]
```
Comment (JordanPawlett) [inline scenes/Lobby.tscn]: Why isn't the handler here `_on_Host_pressed` to conform with the other signal handlers below?

File: scripts/HostButton.gd [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+func _ready():
+	print("Setting IP")
+	var myIP
+	for ip in IP.get_local_addresses():
+		if (not ip.begins_with("127")) and ip.count(".") == 3:
+			print(ip)
+			myIP = ip
+	$MyIP.text = "IP: " + str(myIP)
+
+func host():
```
Comment (JordanPawlett) [inline scripts/HostButton.gd]: See previous comment regarding the naming of this signal handler

File: scripts/Connecting.gd (hunk 2 above)
Comment (wildjames) [reply scripts/Connecting.gd]: Yeah, it's ambiguous. At first I was passing in a string of "client" or "host" before I realised it was dumb. Changed this to a clearer syntax, I think. The function only shows the "Connecting..." screen, so changed the handler to "show_connect_screen(bool is_host)"

File: scripts/HostButton.gd (hunk 3 above)
Comment (wildjames) [reply scripts/HostButton.gd]: The signal triggers the connecting splash screen, rather than setting if it's the host. It sense to separate the setting up of the server/client, and the visuals that get shown to the user. Though, the wording I've used certainly makes this unclear.

Comment (wildjames) [reply scripts/HostButton.gd]: Great idea. Implemented this.

File: scenes/Lobby.tscn (hunk 4 above)
Comment (wildjames) [reply scenes/Lobby.tscn]: No, currently fixed positions. It's definitely something to change, but we need to be careful in a few ways about handling window sizing. 

 - Death only triggers if the player leaves the screen., and walls are a finite height. It's feasible to flap so hard you fly above the walls and never die.

 - We wanna change a few things on resize. By default, making the window bigger just shows you past the "edge", and sprites take up a smaller proportion of the screen.  From extremely brief testing and not having read the docs, fullscreen on my ultrawide vs my laptop makes the bird look huge or tiny. We gotta worry about absolute position on screen, the size of sprites, and the *relative* position between things.

I think this warrants a separate branch to work on, so I won't touch it in this PR. For now, resizing is simply disabled.

File: scenes/Lobby.tscn (hunk 5 above)
Comment (wildjames) [reply scenes/Lobby.tscn]: See above

File: project.godot (hunk 7 above)
Comment (wildjames) [reply project.godot]: A feature we'd probably want in the future would be a centralised server, where players can choose lobbies. If we were to do that, I think this becomes a good point!

File: scripts/Connecting.gd (hunk 8 above)
Comment (wildjames) [reply scripts/Connecting.gd]: Made the structure changes. I'm pretty bad at organising files into sensible directories, so *please* tell me if it's a poor approach. 

Also, somewhere in there is a reference to the `Connecting.gd` script that got broken. Everything seems to work just fine, so I dunno what it's doing, but we do know exactly where it is:

```
 modules/gdscript/gdscript.cpp:829 - Condition "err" is true. Returned: err
 Cannot load source code from file 'res://scripts/Connecting.gd'.
 Failed loading resource: res://scripts/Connecting.gd. Make sure resources have been imported by opening the project in the editor at least once.
```

It's probably fine. Just annoying.

File: scenes/Lobby.tscn (hunk 9 above)
Comment (wildjames) [reply scenes/Lobby.tscn]: Fixed.

File: scripts/HostButton.gd (hunk 10 above)
Comment (wildjames) [reply scripts/HostButton.gd]: Fixed.

**Comment by wildjames** [comment]
Alright, I've been through Jordan's comments, and generally commented the networking code. Should be a bit more readable now, I hope. Also, I fixed the oversight where a solo player is constantly trying to send their position to other players that don't exist. 

Things to consider:
 - Does the `scripts/` structure make sense, and will it continue to make sense as the game increases in complexity?
 - There's a severed link to the `Connecting.gd` script. Easy to fix if we can track it down, but it seems to be in the Godot core `modules` bit and I don't know what it's doing. Error below.


```
 modules/gdscript/gdscript.cpp:829 - Condition "err" is true. Returned: err
 Cannot load source code from file 'res://scripts/Connecting.gd'.
 Failed loading resource: res://scripts/Connecting.gd. Make sure resources have been imported by opening the project in the editor at least once.
```

File: scripts/Netcode/network.gd [hunk 11]
```diff
@@ -0,0 +1,47 @@
+extends Node
+
+const RPC_PORT = 31400
+const MAX_PLAYERS = 2
+const TESTING_IP = "127.0.0.1"
+const OFFLINE_TESTING = true
```
Comment (JordanPawlett) [inline scripts/Netcode/network.gd]: This should be defined by an ENV var

File: scripts/controllers/Lobby.gd [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+		## TODO: Set the random seed here, and broadcast it to all players before the game starts
+		rpc("begin_game")
+		begin_game()
+	else:
+		if not Net.is_host:
+			rpc_id(0, "player_joined")
+
+
+
+remote func player_joined():
```
Comment (JordanPawlett) [inline scripts/controllers/Lobby.gd:20]: It'd be nice to have a player_left method that handled peer_disconnected and decremented the players / cleaned-up the relevant player node

File: scripts/controllers/Lobby.gd [hunk 13]
```diff
@@ -0,0 +1,29 @@
+extends Control
+
+
+func _ready():
+	var _junk = get_tree().connect("connected_to_server", self, "connected")
```
Comment (JordanPawlett) [inline scripts/controllers/Lobby.gd:5]: I believe the server should be listening to
`network_peer_connected` and `network_peer_disconnected` to handle players joining and leaving.

`connected_to_server` is a client event, so currently the client is in charge of incrementing the players.

File: scripts/UI/Connecting.gd [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+	# 1: Client
+	# 2: Spectator ## TODO
+	# 
+	# Should allow for some degree of expansion.
+	
+	
+	# Reveal thyself
+	show()
+	# Host
+	if conn_type == 0:
```
Comment (Jibbajabbafic) [inline scripts/UI/Connecting.gd]: This would be better off as an enum probably. Would function identically but makes it much more readable e.g.
```
enum ConnType {HOST, CLIENT, SPECTATOR}
# ^ this line would probably be in some other file so Connecting.gd and HostButton.gd can make use of it, idk if godot can handle enums being defined in their own file
if conn_type == ConnType.HOST:
    ...
```

File: scripts/controllers/World.gd [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+
+func _on_WallSpawner_timeout():
+	WallSpawnTimer.wait_time = wall_spawn_time * float(start_wall_speed) / wall_speed
+	spawn_wall()
+
+
+#### Player helper functions
+func _on_Player_death(player):
+	# See if we have a new PB
+	if player.score > int(HiScore.text):
```
Comment (Jibbajabbafic) [inline scripts/controllers/World.gd]: For UI you should always keep the values as variables and only use the label to display it, makes it much easier to work with so you don't have to keep grabbing values from it

File: scripts/controllers/World.gd [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+	# Actual incrimenting is handled on the player object
+	increase_difficulty()
+	Score.text = str(player.score)
+
+
+#### World functions
+func increase_difficulty():
+	wall_speed += speed_up
+	# Speed up the background
+	# Background.material.set_shader_param('scroll_speed', wall_speed*parallax)
```
Comment (Jibbajabbafic) [inline scripts/controllers/World.gd:96]: With commented code if it ain't being used it's usually better to just remove it, or if you do still need it add a TODO to explain why

File: scripts/controllers/World.gd [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+	return high_score
+
+func save_score(score):
+	var saved = File.new()
+	saved.open(high_score_fname, File.WRITE)
+	
+	# We will store as a JSON object. Overkill for a single integer, but should 
+	# be easy to scale out.
+	var score_dict = {"highscore": score}
+	saved.store_line(to_json(score_dict))
```
Comment (Jibbajabbafic) [inline scripts/controllers/World.gd]: You should probably close the file after writing to make sure no nastyness happens.I.e.
```suggestion
	saved.store_line(to_json(score_dict))
	saved.close()
```
The name `saved` feels slightly confusing to read too so something like `save_file` would probably work better but that's a bit of a nitpick

File: scripts/controllers/World.gd [hunk 18]
```diff
@@ -0,0 +1,119 @@
+extends Node2D
```
Comment (Jibbajabbafic) [inline scripts/controllers/World.gd:1]: In future when renaming files it's better to do that in a separate commit/PR so git can actually track the rename properly. When you make changes _and_ rename I think it confuses github and you lose all of the history, so it's much harder to review what's changed

File: scripts/controllers/World.gd [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+var wall_speed = start_wall_speed
+var speed_up = 0.1
+var scroll_speed = 0.1
+var bob_speed = 1.0
+var bob_amplitude = 0.01
+
+
+func _ready():
+	# Setup gubbins
+	HiScore.text = str(load_score())
```
Comment (Jibbajabbafic) [inline scripts/controllers/World.gd]: When testing I think it kept reloading the save file every time you die because the scene gets reloaded, so it might be better to load this when the game first starts up somewhere, but this is fine for now, just something to be aware of later on

File: scripts/controllers/World.gd [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+func load_score():
+	var high_score = 0
+	
+	var saved = File.new()
+	if not saved.file_exists(high_score_fname):
+		return high_score
+	
+	saved.open(high_score_fname, File.READ)
+	while saved.get_position() < saved.get_len():
+		var data = parse_json(saved.get_line())
```
Comment (Jibbajabbafic) [inline scripts/controllers/World.gd]: I'm not really familiar with the file handler but parsing JSON line by line looks weird to me , I would have thought the whole file could be parsed as one JSON object

File: scripts/controllers/World.gd [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+	# Background.material.set_shader_param('scroll_speed', wall_speed*parallax)
+	for wall in get_tree().get_nodes_in_group("walls"):
+		wall.speed = wall_speed
+
+func load_score():
+	var high_score = 0
+	
+	var saved = File.new()
+	if not saved.file_exists(high_score_fname):
+		return high_score
```
Comment (Jibbajabbafic) [inline scripts/controllers/World.gd]: This seems kinda misleading, I think just returning 0 would be a lot clearer that no highscore exists

File: shaders/scroll_and_bob.shader [hunk 22]
```diff
@@ -0,0 +1,22 @@
...
+	
+	// This simply takes the time run, multiplies the speed by it, and sets the 
+	// new x to that.
+	// This means that altering the speed results in jumps. Need to fix. 
+	// Can shaders access dt?
+	shiftedUV.x += TIME * scroll_speed;
+	shiftedUV.y += bob_amplitude * sin(TIME*bob_speed);
+	
+	vec4 col = texture(TEXTURE, shiftedUV);
+
```
Comment (Jibbajabbafic) [inline shaders/scroll_and_bob.shader:20]: I think it's awesome you made a shader, they seem like black magic to me

File: sprites/frog_pond.jpg.import [hunk 23]
```diff
@@ -0,0 +1,34 @@
+[remap]
```
Comment (Jibbajabbafic) [inline sprites/frog_pond.jpg.import:1]: lol frog pond

File: scripts/UI/HostButton.gd [hunk 24]
```diff
@@ -0,0 +1,20 @@
+extends Node
+
+signal show_connect_screen
+
+# Called when the button is loaded.
+func _ready():
+	var myIP
+	# TODO: Is this reliable, and portable between OS?
+	for ip in IP.get_local_addresses():
```
Comment (Jibbajabbafic) [inline scripts/UI/HostButton.gd:9]: Seeing similar logic in this script and Connecting.gd, is there some way to keep all the logic together and then just tie the buttons to it? Seems like that would be cleaner than a script for each button, but idk if that's possible so I could be wrong

**Review by Jibbajabbafic — COMMENTED** [review]
Here's some feedback, some of it might be a bit nitpicky tho so don't worry about all of it, just good to bear in mind for future code

File: scripts/UI/HostButton.gd (hunk 24 above)
Comment (wildjames) [reply scripts/UI/HostButton.gd:9]: Done. Moved to the network.gd script, which is globally accessible

File: sprites/frog_pond.jpg.import (hunk 23 above)
Comment (wildjames) [reply sprites/frog_pond.jpg.import:1]: lol frog pond

File: scripts/controllers/World.gd (hunk 20 above)
Comment (wildjames) [reply scripts/controllers/World.gd]: Made this change

File: scripts/controllers/World.gd (hunk 19 above)
Comment (wildjames) [reply scripts/controllers/World.gd]: Created a Globals.gd script, that's accessible from any node. Currently only stores the high score, but can have general useful stuff added to it later.

File: scripts/UI/Connecting.gd (hunk 14 above)
Comment (wildjames) [reply scripts/UI/Connecting.gd]: Done

File: scripts/controllers/World.gd [hunk 25]
```diff
@@ -0,0 +1,122 @@
...
+	for wall in get_tree().get_nodes_in_group("walls"):
+		wall.speed = wall_speed
+
+func load_high_score():
+	var save_file = File.new()
+	if not save_file.file_exists(high_score_fname):
+		return 0
+	
+	save_file.open(high_score_fname, File.READ)
+	var data = parse_json(save_file.get_line())
```
Comment (Jibbajabbafic) [inline scripts/controllers/World.gd]: You can parse the whole file as JSON with this
```suggestion
	var data = parse_json(save_file.get_as_text())
```
